monitor/src/main/java/edu/asu/diging/monitor/web/admin/ModifyGroupController.java

```diff
@@ -0,0 +1,115 @@
+        binder.addValidators(groupValidator);
+    }
+
+    @RequestMapping(value = "/admin/groups/{id}/modify", method = RequestMethod.GET)
+    public String show(Model model, @PathVariable("id") String id, RedirectAttributes redirectAttrs) {
+        GroupForm groupForm = new GroupForm();
+        groupForm.setApps(appManager.getApps().stream().map(a -> {
+            AppForm app = new AppForm();
+            app.setId(a.getId());
+            app.setName(a.getName());
+            return app;
+        }).collect(Collectors.toList()));
+        Group group = null;
+        try {
+            group = groupManager.getGroup(id);
+        } catch (GroupNotFoundException e) {
+            logger.error("Could not modify group", e);
+            redirectAttrs.addFlashAttribute("show_alert", true);
+            redirectAttrs.addFlashAttribute("alert_type", "danger");
+            redirectAttrs.addFlashAttribute("alert_msg", "The selected group doesn't exist.");
+            return "redirect:/admin/groups/show";
+        }
+        groupForm.setId(id);
+        groupForm.setName(group.getName());
+        model.addAttribute("groupForm", groupForm);
+        model.addAttribute("appIds", group.getApps().stream().map(a -> a.getId()).collect(Collectors.toList()));
+        return "admin/groups/add";
+    }
+
+    @RequestMapping(value = "/admin/groups/{id}/modify", method = RequestMethod.POST)
+    public String add(@ModelAttribute @Validated GroupForm groupForm, BindingResult result,
+            RedirectAttributes redirectAttrs, @PathVariable("id") String id) {
+        if (result.hasErrors()) {
+            redirectAttrs.addFlashAttribute("show_alert", true);
+            redirectAttrs.addFlashAttribute("alert_type", "danger");
+            redirectAttrs.addFlashAttribute("alert_msg", result.getFieldError().getCode());
+            return "redirect:/admin/groups/{id}/modify";
+        }
+        Group group = null;
+        try {
+
+            group = groupManager.getGroup(id);
+
+        } catch (GroupNotFoundException e) {
+            logger.error("Could not update group.", e);
+            redirectAttrs.addFlashAttribute("show_alert", true);
+            redirectAttrs.addFlashAttribute("alert_type", "danger");
+            redirectAttrs.addFlashAttribute("alert_msg", "Group does not exist");
+            return "redirect:/";
+        }
+        if (!group.getName().equals(groupForm.getName())) {
+            group.setName(groupForm.getName());
+            groupManager.updateGroup(group);
+        }
+        group.getApps().stream().forEach(a -> {
+            a.setGroup(null);
+            appManager.updateApp(a);
+        });
+        Group updatedGroup = group;
+        if (groupForm.getAppIds() != null) {
+            groupForm.getAppIds().stream().map(appId -> appManager.getApp(appId)).forEach(a -> {
+                a.setGroup(updatedGroup);
+                appManager.updateApp(a);
+            });
+        }
```

Comment:
all of the above (starting line 94) is business logic and should go into a manager class.

Reply:
business logic